### PR TITLE
Multi-Tier database backend for long term metrics storage

### PR DESCRIPTION
**Pull request by stelfrag**
## Netdata database tiering support

This PR allows a Netdata agent to use multiple dbengines concurrently, one for a separate tier of data aggregation.

Abstract from a netdata.conf:

```
[db]
	storage tiers = 3

	dbengine page cache size MB = 32
	dbengine multihost disk space MB = 256

	dbengine tier 1 page cache size MB = 32
	dbengine tier 1 multihost disk space MB = 256
	dbengine tier 1 update every iterations = 60

	dbengine tier 2 page cache size MB = 32
	dbengine tier 2 multihost disk space MB = 256
	dbengine tier 2 update every iterations = 60
```

The maximum number of tiers is currently 5 (it can be increased at compile time - currently some constructors at needed). At run time, users may choose any number of tiers, from 1 to the compiled maximum.

Tier 0 is the default that was always available. Tier 1 is the first level of aggregation. Tier 2 is the second.

Metrics on all tiers > 1 store 5 values for every point:

1. The `sum` of the points aggregated
2. The `min` of the points aggregated
3. The `max` of the points aggregated
4. The `count` of the points aggregated (could be constant, but it may not be due to gaps in data collection)
5. The `anomaly_count` of the points aggregated (how many of the aggregated points found anomalous)

Among `min`, `max` and `sum`, the right value is chosen based on the user query. `average` is calculated on the fly at query time.

### storage

This PR introduces a new page type for dbengine. Instead of storing packed numbers in it, we store unpacked all the values as a single number 16 bytes long (3x float + 2x uint16). So, the maximum number of points that can be aggregated is 65535 to any tier.

The numbers are pushed to each tier during data collection in real-time - no additional threads are introduced. Each tier is aligned to wall time. So if tier 0 has 1s granularity, and tier 1 is set to every 60 iterations, metrics will stored at XX:XX:00 of every minute. Similarly if tier 0 has 2s granularity, and tier 1 is set to every 60 iterations, metrics will be stored at 00 seconds on even minutes.

### querying

The query engine selects automatically the right tier to query based on the following:

1. ~It calculates the time coverage % of each tier for the timeframe requested (does it have data for the entire duration requested?).~ no longer needed - the query engine always uses all tier to satisfy a query
2. It calculates the points coverage % of each tier for the points requested (does it have the required resolution to provide the desired number of points?).
3. ~It sums the two percentages as the final score of each tier.~ no longer needed - time coverage is guaranteed now.
4. Picks the tier with biggest score for the query. If many tiers have the same score, the highest tier among them is preferred.

The query engine has been re-written to make the best use of higher tiers that contain less points. Before this PR the internal heartbeat of the query engine was `update_every` (1s for virtual points). Now its heartbeat is `view_update_every` and it automatically handles high and low resolution databases. In case the database resolution is too low for a query, interpolation kicks in to generate points at `view_update_every` resoluytion.

---

### TODO

- [x] ~the query engine when it comes to virtual points has a fixed granularity of 1 second. So, when choosing a higher tier, although the points will be less, the query will actually do a loop at 1s granularity. It will be faster than fetching points, but it needs to be changed to take all the value from higher tiers.~
- [x] ~the query engine should use all the available tier to fulfill the request, so that if data are available to any other tier, it should switch tiers on the fly to satisfy the request~

### CODE REVIEW

File: database/engine/rrdengineapi.c [hunk 1]
```diff
@@ -195,20 +216,44 @@ void rrdeng_store_metric_flush_current_page(RRDDIM *rd)
...
+    RRDDIM *rd = metric_handle->rd;
+    int tier = ctx->tier;
 
-    ctx = handle->ctx;
-    pg_cache = &ctx->pg_cache;
-    descr = handle->descr;
+    storage_number number;
+    storage_number_tier1_t number_tier1;
+
+    number = pack_storage_number(n, flags);
```
Comment (MrZammler) [inline database/engine/rrdengineapi.c:241]: Hi! Is this needed here, since it is done again below ?

File: database/rrdhost.c [hunk 2]
```diff
@@ -746,24 +825,100 @@ int rrd_init(char *hostname, struct rrdhost_system_info *system_info) {
...
+            if(grouping_iterations < 2) {
+                grouping_iterations = 2;
+                config_set_number(CONFIG_SECTION_DB, dbengineconfig, grouping_iterations);
+                error("DBENGINE on '%s': 'dbegnine tier %d update every iterations' cannot be less than 2. Assuming 2.", localhost->hostname, tier);
+            }
+            if(grouping_iterations > 255) {
+                grouping_iterations = 255;
+                config_set_number(CONFIG_SECTION_DB, dbengineconfig, grouping_iterations);
+                error("DBENGINE on '%s': 'dbegnine tier %d update every iterations' cannot be more than the previous tier. Assuming 255.", localhost->hostname, tier);
+            }
```
Comment (MrZammler) [inline database/rrdhost.c:866]: Hi, not sure of the above check, what it's trying to do... ?

**Comment by vlvkobal** [comment]
Valgrind sometimes reports lost memory blocks on Netdata exit
```
==196205== 256 bytes in 8 blocks are definitely lost in loss record 691 of 763
==196205==    at 0x4845888: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==196205==    by 0x1482D3: mallocz (libnetdata.c:170)
==196205==    by 0x25B89B: rrdeng_metric_init (rrdengineapi.c:139)
==196205==    by 0x19F53E: create_rrdim_entry (sqlite_functions.c:1609)
==196205==    by 0x1A3F54: sql_build_context_param_list (sqlite_functions.c:1716)
==196205==    by 0x16C83B: web_client_api_request_v1_data (web_api_v1.c:586)
==196205==    by 0x16D9E0: web_client_api_request_v1 (web_api_v1.c:1451)
==196205==    by 0x27384B: web_client_api_request (web_client.c:537)
==196205==    by 0x273B8E: check_host_and_call (web_client.c:513)
==196205==    by 0x273B8E: web_client_process_url (web_client.c:1359)
==196205==    by 0x275362: web_client_process_request (web_client.c:1505)
==196205==    by 0x277D76: web_server_rcv_callback (static-threaded.c:304)
==196205==    by 0x152E5F: poll_process_tcp_read (socket.c:1440)
```

**Comment by vlvkobal** [comment]
It looks like this part doesn't makes `points_wanted` zero
```
  if(automatic_natural_points) {
        points_wanted = (before_wanted - after_wanted + 1) / query_granularity;
        query_debug_log(":auto natural points_wanted %ld", points_wanted);
    }
```

File: database/rrdhost.c (hunk 2 above)
Comment (ktsaou) [reply database/rrdhost.c:866]: It is a left-over. We were thinking to limit the max number of iterations to 255 (a byte) to make the counts inside the tier1 and tier2 from uint16_t and uint8_t but we dropped this idea. Removing this check.

**Comment by ktsaou** [comment]
> Valgrind sometimes reports lost memory blocks on Netdata exit

This is very strange. I merged the freeing of this code with the freeing of `rd->tiers[]`, so please check it again.

**Comment by ktsaou** [comment]
Guys, we addressed all your comments.
Please give it another try.

File: database/engine/rrdengineapi.c (hunk 1 above)
Comment (ktsaou) [reply database/engine/rrdengineapi.c]: removed it

**Comment by vlvkobal** [comment]
There is a problem with `rrddim_query_next_metric()`.
It is `extern NETDATA_DOUBLE rrddim_query_next_metric(struct rrddim_query_handle *handle, time_t *start_time, time_t *end_time, SN_FLAGS *flags)` in master but it is `extern STORAGE_POINT rrddim_query_next_metric(struct rrddim_query_handle *handle)` in this PR.

**Comment by vlvkobal** [comment]
Have we changed NETDATA_DOUBLE again?

**Comment by thiagoftsm** [comment]
@ktsaou and @stelfrag I am still reviewing the code, and I decided to start some tests to speed up the process. When I start netdata using `memory mode = ram` I receive these errors:

```sh
2022-07-05 16:39:04: netdata FATAL : netdata : (0012@database/rrdset.c   :__rrdset_check_): RRDSET 'apps.sockets' should be read-locked, but it is not, at function rrdset_last_entry_t_nolock() at line 1175 of file './database/rrd.h' # : Success

2022-07-05 16:39:04: netdata LOG FLOOD PROTECTION reset for process 'netdata' (prevented 323 logs in the last 9 seconds).
2022-07-05 16:39:04: netdata INFO  : MAIN : (0031@daemon/main.c       :netdata_cleanup): EXIT: netdata prepares to exit with code 1...
2022-07-05 16:39:04: netdata INFO  : MAIN : (0042@daemon/main.c       :netdata_cleanup): EXIT: cleaning up the database...
2022-07-05 16:39:04: netdata INFO  : MAIN : (1418@database/rrdhost.c  :rrdhost_cleanup): Cleaning up database [1 hosts(s)]...
2022-07-05 16:39:04: netdata INFO  : MAIN : (1373@database/rrdhost.c  :rrdhost_cleanup): Cleaning up database of host 'hades'...
2022-07-05 16:39:04: netdata ERROR : MAIN : (0223@libnetdata/locks/loc:__netdata_rwloc): RW_LOCK: failed to obtain read lock (code 35)
2022-07-05 16:39:04: netdata INFO  : MAIN : (0504@database/sqlite/sqli:sql_close_datab): Closing SQLite database
2022-07-05 16:39:04: netdata INFO  : MAIN : (0105@database/sqlite/sqli:add_stmt_to_lis): No statements pending to finalize
2022-07-05 16:39:04: netdata INFO  : MAIN : (0079@daemon/main.c       :netdata_cleanup): EXIT: all done - netdata is now exiting - bye bye...
EOF found in spawn pipe.
Shutting down spawn server event loop.
Shutting down spawn server loop complete.

```

When I am using `dbengine` everything works as expected.

**Comment by thiagoftsm** [comment]
I have analogous [issues](https://github.com/netdata/netdata/pull/13263#issuecomment-1175272076) with `map`, and `save`.

File: database/engine/rrdengineapi.c [hunk 3]
```diff
@@ -977,7 +871,7 @@ int rrdeng_init(RRDHOST *host, struct rrdengine_instance **ctxp, char *dbfiles_p
     finalize_rrd_files(ctx);
 error_after_init_rrd_files:
     free_page_cache(ctx);
-    if (ctx != &multidb_ctx) {
+    if (!is_storage_engine_shared((STORAGE_INSTANCE *)ctx)) {
```
[the remark this answers is not in the document]
Comment (ktsaou) [reply database/engine/rrdengineapi.c:878]: should be fixed now

**Comment by vkalintiris** [comment]
AFAIC, it's fine to merge this with the given default configuration values. A couple of (semi-random) things to note:

- It would be nice to have a global `enable experimental|alpha|beta|<whatever appropriate> tiering feature = yes|no` and enable tiering only when this is set. The point is to make it crystal clear to the users that a feature is not stable yet.
- IIUC, we are only flushing a page when it get's full. A 3rd level tier page getting a new value every 60 * 60 seconds, will end up keeping in RAM data for a _really_, _really_ _long_ time-range (esp. if we double/quadruple the page size). Getting a crash, OOM or an out-of-disk situation can cause extensive data loss.
- CPU-wise I only noticed a negligible CPU increase when (a) enabling multi-tier support, and (b) stress testing with 10000 charts having 10 dims each (from a single-thread collector), ie. not anything to justify delving into at this point. Curious, how CPU and I/O will look when we need to flush full pages at once from all tiers.

**Comment by ktsaou** [comment]
@thiagoftsm please check the db modes `ram`, `map`, `save` and `alloc` again.

**Comment by MrZammler** [comment]
Got this alert during a test (only once, can't reproduce):

image

Nothing really during that time in logs about DBENGINE, close to a start I believe. Backfilling for this chart and dimension happened ~6 minutes after the alert, but shouldn't matter, since health plays from tier 0... It seems as if it couldn't find the dimension, or values there...

Also got this core from a query:

```
Core was generated by `/home/evas/stuff/netdata/install_netdata_tiering/netdata/usr/sbin/netdata -D -W'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x00005620b02e410b in rrdeng_metric_oldest_time (db_metric_handle=0x5620b47666c0)
    at database/engine/rrdengineapi.c:588
588	   return page_index->oldest_time / USEC_PER_SEC;
[Current thread is 1 (Thread 0x7f73a36ed640 (LWP 22408))]
(gdb) where
#0  0x00005620b02e410b in rrdeng_metric_oldest_time (db_metric_handle=0x5620b47666c0)
    at database/engine/rrdengineapi.c:588
#1  0x00005620b01fd947 in rrddim_first_entry_t (rd=<optimized out>) at ./database/rrd.h:1162
#2  rrdset_first_entry_t_nolock (st=0x5620b47644c0) at ./database/rrd.h:1199
#3  build_context_param_list (owa=owa@entry=0x5620b4b8d0c0, param_list=param_list@entry=0x7f73a36ec9a8, 
    st=st@entry=0x5620b47644c0) at web/api/formatters/rrd2json.c:95
#4  0x00005620b0205f8a in web_client_api_request_v1_data (host=0x5620b37484d0, w=0x5620b4b7af10, 
    url=<optimized out>) at web/api/web_api_v1.c:568
#5  0x00005620b0207b6d in web_client_api_request_v1 (host=host@entry=0x5620b37484d0, 
    w=w@entry=0x5620b4b7af10, url=url@entry=0x5620b491e922 "data") at web/api/web_api_v1.c:1451
#6  0x00005620b0305ca9 in aclk_web_api_v1_request (url=0x5620b491e922 "data", w=0x5620b4b7af10, 
    host=0x5620b37484d0) at aclk/aclk_query.c:21
#7  http_api_v2 (query=0x5620b4b703f0, query_thr=0x5620b3df5cb0) at aclk/aclk_query.c:147
#8  aclk_query_process_msg (query=0x5620b4b703f0, query_thr=0x5620b3df5cb0) at aclk/aclk_query.c:297
#9  aclk_query_process_msgs (query_thr=query_thr@entry=0x5620b3df5cb0) at aclk/aclk_query.c:322
#10 0x00005620b03066ab in aclk_query_main_thread (ptr=0x5620b3df5cb0) at aclk/aclk_query.c:344
#11 0x00005620b01f1665 in thread_start (ptr=<optimized out>) at libnetdata/threads/threads.c:185
#12 0x00007f73bf51eef7 in start_thread () from /lib64/libc.so.6
#13 0x00007f73bf5a333c in clone3 () from /lib64/libc.so.6
```

The query executed was: `/node/d7ab9796-b134-494b-8754-49697bc4493c/api/v1/data?after=1657044128&before=1657045028&context=system.cpu&dimensions=&format=json&group=average&gtime=0&options=absolute%7Callow_past%7Cflip%7Cjsonwrap%7Cms%7Cpercentage&points=441&show_dimensions=1&timeout=9000`

**Comment by stelfrag** [comment]
@MrZammler trying to reproduce

**Comment by thiagoftsm** [comment]
@ktsaou now all db modes are working when we compile this PR, I will proceed with tests and code review.

**Comment by thiagoftsm** [comment]
@ktsaou and @stelfrag I am running the [script](https://github.com/netdata/netdata/tree/master/tests/urls) with all memory modes to identify possible issues with API, and I observed that `netdata` crashes when I access next URL using `mode = ram`:

```sh
$ curl https://192.168.0.50:19999/api/v1/badge.svg?chart=apps.bandwidth_tcp_recv&points=100&before=1657049743&after=1657049773&group=median
```

The last messages before the crash are:

```sh
2022-07-05 19:38:17: netdata INFO  : MAIN : (0144@database/sqlite/sqli:prepare_stateme): Thread 31781: Using key 4 on statement 0x7820118
double free or corruption (out)
EOF found in spawn pipe.
Shutting down spawn server event loop.
Shutting down spawn server loop complete.
```

The backtrace of the event:

```sh
Thread 38 "WEB_SERVER[stat" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fffe391c640 (LWP 314)]
0x00007ffff6e9d617 in _int_free () from /lib64/libc.so.6
(gdb) bt
#0  0x00007ffff6e9d617 in _int_free () from /lib64/libc.so.6
#1  0x00007ffff6e9fe7f in free () from /lib64/libc.so.6
#2  0x000000000043d111 in freez (ptr=<optimized out>) at libnetdata/libnetdata.c:166
#3  0x00000000004557da in grouping_free_median (r=0x7079740) at web/api/queries/median/median.c:38
#4  0x0000000000458e14 in rrd2rrdr (owa=<optimized out>, st=<optimized out>, st@entry=0x614d9a0, points_requested=points_requested@entry=100, after_requested=after_requested@entry=1657049773, 
    before_requested=before_requested@entry=1657049743, group_method=group_method@entry=RRDR_GROUPING_MEDIAN, resampling_time_requested=<optimized out>, options=<optimized out>, dimensions=<optimized out>, 
    context_param_list=<optimized out>, group_options=<optimized out>, timeout=<optimized out>, tier=<optimized out>) at web/api/queries/query.c:1727
#5  0x000000000045a648 in rrdset2value_api_v1 (st=st@entry=0x614d9a0, wb=<optimized out>, n=n@entry=0x7fffe3919fe0, dimensions=<optimized out>, points=points@entry=100, after=after@entry=1657049773, before=<optimized out>, 
    group_method=<optimized out>, group_options=<optimized out>, group_time=<optimized out>, options=<optimized out>, db_after=<optimized out>, db_before=<optimized out>, db_points_read=<optimized out>, 
    result_points_generated=<optimized out>, value_is_null=<optimized out>, anomaly_rate=<optimized out>, timeout=<optimized out>, tier=<optimized out>) at web/api/formatters/rrd2json.c:192
#6  0x00000000004536e3 in web_client_api_request_v1_badge (host=<optimized out>, w=0x70c9030, url=<optimized out>) at web/api/badges/web_buffer_svg.c:1109
#7  0x0000000000464885 in web_client_api_request_v1 (host=host@entry=0x24f33f0, w=w@entry=0x70c9030, url=0x70c9cc6 "badge.svg") at web/api/web_api_v1.c:1451
#8  0x000000000066410c in web_client_api_request (host=host@entry=0x24f33f0, w=w@entry=0x70c9030, url=<optimized out>) at web/server/web_client.c:537
#9  0x00000000006644f8 in check_host_and_call (func=0x66409a <web_client_api_request>, url=<optimized out>, w=0x70c9030, host=0x24f33f0) at web/server/web_client.c:513
#10 web_client_process_url (host=0x24f33f0, w=w@entry=0x70c9030, url=<optimized out>, url@entry=0x70c9cbd "") at web/server/web_client.c:1359
#11 0x00000000006661b6 in web_client_process_request (w=w@entry=0x70c9030) at web/server/web_client.c:1505
#12 0x0000000000668caa in web_server_rcv_callback (pi=0x7012d60, events=0x702b744) at web/server/static/static-threaded.c:304
#13 0x0000000000449d2b in poll_process_tcp_read (now=2570, pf=0x702b740, pi=0x7012d60, p=0x7fffe391b410) at libnetdata/socket/socket.c:1440
#14 poll_events (sockets=<optimized out>, add_callback=add_callback@entry=0x6695aa <web_server_add_callback>, del_callback=del_callback@entry=0x669216 <web_server_del_callback>, 
    rcv_callback=rcv_callback@entry=0x668c29 <web_server_rcv_callback>, snd_callback=snd_callback@entry=0x668a7e <web_server_snd_callback>, tmr_callback=tmr_callback@entry=0x0, access_list=<optimized out>, allow_dns=<optimized out>, 
    data=<optimized out>, tcp_request_timeout_seconds=<optimized out>, tcp_idle_timeout_seconds=<optimized out>, timer_milliseconds=<optimized out>, timer_data=<optimized out>, max_tcp_sockets=<optimized out>)
    at libnetdata/socket/socket.c:1759
#15 0x0000000000668a35 in socket_listen_main_static_threaded_worker (ptr=<optimized out>) at web/server/static/static-threaded.c:427
#16 0x000000000044c278 in thread_start (ptr=<optimized out>) at libnetdata/threads/threads.c:185
#17 0x00007ffff6e90053 in start_thread () from /lib64/libc.so.6
#18 0x00007ffff6f1cbf0 in clone () from /lib64/libc.so.6

```

**Comment by ktsaou** [comment]
You are amazing guys!

All issues fixed, please review / test again.

**Comment by thiagoftsm** [comment]
Backward compatibility worked as expected:

| Mode | Test | Tier |
|------|------|------|
| save | API  | no |
| save | Stremaing  | no |
| save | Health  | no |
| save | Cloud | no |
| save | Plugins | no |
| save | netdatacli | no |
| ram | API  | no |
| ram | Stremaing  | no |
| ram | Health  | no |
| ram | Cloud | no |
| ram | Plugins | no |
| ram | netdatacli | no |
| map | API  | no |
| map | Stremaing  | no |
| map | Health  | no |
| map | Cloud | no |
| map | Plugins | no |
| map | netdatacli | no |

Testing now the new feature.

**Review by thiagoftsm — APPROVED** [review]
After hours testing on my environment, I had netdata working with [current](https://github.com/netdata/netdata/pull/13263#issuecomment-1175548373) PR''s code, and I also ran it with Tier enabled. On parent I had the following setup:

```sh
storage tiers = 3

        dbengine tier 1 page cache size MB = 32
        dbengine tier 1 multihost disk space MB = 256
        dbengine tier 1 update every iterations = 5

        dbengine tier 2 page cache size MB = 32
        dbengine tier 2 multihost disk space MB = 256
        dbengine tier 2 update every iterations = 5
```

and on child I ran with the following:

```sh
storage tiers = 3

        dbengine tier 1 page cache size MB = 32
        dbengine tier 1 multihost disk space MB = 256
        dbengine tier 1 update every iterations = 3

        dbengine tier 2 page cache size MB = 32
        dbengine tier 2 multihost disk space MB = 256
        dbengine tier 2 update every iterations = 5
```

After this I tested different settings:

| Mode | Test | Tier |
|------|------|------|
| save | API  | yes |
| save | API specific Tier | yes |
| save | API Tier not available | yes |
| save | Stremaing  | yes |
| save | Health  | no |
| save | Cloud | yes |
| save | Plugins | yes |
| save | netdatacli | yes |
| ram | API  | yes |
| ram | Child API  | yes |
| ram | Stremaing  | yes |
| ram | Health  | yes |
| ram | Cloud | yes |
| ram | Plugins | yes |
| ram | netdatacli | yes |
| map | API  | yes |
| map | Stremaing  | yes |
| map | Health  | yes |
| map | Cloud | yes |
| map | Plugins | yes |
| map | netdatacli | yes |
| dbengine | API  | yes |
| dbengine | Stremaing  | yes |
| dbengine | Health  | yes |
| dbengine | Cloud | yes |
| dbengine | Plugins | yes |
| dbengine | Health  | yes |

And everything worked as expected. Finally I compile our current master branch and I could have everything running as expected. LGTM!

**Comment by MrZammler** [comment]
Latest updates seem to have fixed the crashes which would otherwise be reproducible.

Even though cloud (prod and staging) seemed a little finicky (missing initial start streaming commands) it seems now to work okay! So it could have been something this morning from the cloud side.

I got the same alert again (10min_dbengine_global_flushing_errors) we'll need to investigate that.

File: database/engine/rrdengineapi.c [hunk 4]
```diff
@@ -603,21 +469,21 @@ static int rrdeng_load_page_next(struct rrddim_query_handle *rrdimm_handle) {
 // Returns the metric and sets its timestamp into current_time
 // IT IS REQUIRED TO **ALWAYS** SET ALL RETURN VALUES (current_time, end_time, flags)
 // IT IS REQUIRED TO **ALWAYS** KEEP TRACK OF TIME, EVEN OUTSIDE THE DATABASE BOUNDARIES
-NETDATA_DOUBLE
-rrdeng_load_metric_next(struct rrddim_query_handle *rrdimm_handle, time_t *start_time, time_t *end_time, SN_FLAGS *flags) {
+STORAGE_POINT rrdeng_load_metric_next(struct rrddim_query_handle *rrdimm_handle) {
```
Comment (vkalintiris) [inline database/engine/rrdengineapi.c:472]: I'm curious if returning STORAGE_POINT will hurt performance significantly. I haven't checked x86_64's ABI, but my guess is that `STORAGE_POINT` values can't get materialized in registers.

Comment (ktsaou) [reply database/engine/rrdengineapi.c:472]: I guess this happens through the stack. Right?
I have not tested it and it is a really simple change to pass it as a pointer to the function, but then the function will have to dereference it for each of its members. Can you test the difference?

**Pull request merged**